examples/usdt/main.go

```diff
@@ -0,0 +1,403 @@
+	fmt.Printf("%s\n\n", note.String())
+
+	// Open an ELF binary.
+	ex, err := link.OpenExecutable(note.path)
+	if err != nil {
+		log.Fatalf("open executable: %v", err)
+	}
+
+	// Open a Uprobe on the SDT note address and provider path.
+	sym := fmt.Sprintf("ebpf_usdt_%s_%s", provider, probe)
+	opts := &link.UprobeOptions{
+		PID:          *pid,
+		Offset:       note.location,
+		RefCtrOffset: note.semaphore_offset_refctr,
+	}
+	u, err := ex.Uprobe(sym, objs.Handler, opts)
+	if err != nil {
+		if !errors.Is(err, link.ErrNotSupported) {
+			log.Fatalf("create uprobe (with ref_ctr_offset): %v", err)
+		}
+
+		// Fallback to manual semaphore handling (Kernel 4.20-)
+		opts.RefCtrOffset = 0
+		u, err = ex.Uprobe(sym, objs.Handler, opts)
+		if err != nil {
+			log.Fatalf("create uprobe: %v", err)
+		}
+		if note.semaphore != 0 {
+			if err := semUpdate(*pid, note.semaphore_offset_ptrace, note.bo, true); err != nil {
+				log.Fatalf("inc semaphore: %v", err)
+			}
+			defer func() {
+				if err := semUpdate(*pid, note.semaphore_offset_ptrace, note.bo, false); err != nil {
+					log.Fatalf("dec semaphore: %v", err)
+				}
+			}()
+		}
+	}
+	defer u.Close()
+
+	// Open a ringbuf reader from userspace RINGBUF map.
+	rd, err := ringbuf.NewReader(objs.Events)
+	if err != nil {
+		log.Fatalf("opening ringbuf reader: %v", err)
+	}
+	defer rd.Close()
+
+	// Close the reader when the process receives a signal, which will exit
+	// the read loop.
+	go func() {
+		<-stopper
+		if err := rd.Close(); err != nil {
+			log.Fatal(err)
+		}
+	}()
+
+	fmt.Printf("FILENAME\t\t\t\tFUNCTION\n")
+	for {
+		record, err := rd.Read()
+		if err != nil {
+			if errors.Is(err, ringbuf.ErrClosed) {
+				log.Println("received signal, exiting..")
+				return
+			}
+			log.Printf("read from reader: %v", err)
+			continue
+		}
+
+		fmt.Printf(
+			"%s\t\t%s\n",
+			string(record.RawSample[:strsz]),
+			string(record.RawSample[strsz:]),
+		)
+	}
+}
+
+// Read /proc/<pid>/maps and search for the desired note.
+//
+// Skip reading notes in the Python executable as we know for sure they
+// will be provided by a shared object.
+func loadNote(pid int, provider, probe string) (*sdtNoteInfo, error) {
+	m, err := os.Open(fmt.Sprintf("/proc/%d/maps", pid))
+	if err != nil {
+		return nil, err
+	}
+	defer m.Close()
+
+	s := bufio.NewScanner(m)
+	seen := make(map[string]struct{})
+	for s.Scan() {
+		ss := strings.Split(s.Text(), " ")
+
+		path := ss[len(ss)-1]
+		if _, ok := seen[path]; ok {
+			continue
+		}
+
+		addr := strings.Split(ss[0], "-")
+		start, err := strconv.ParseUint(addr[0], 16, 64)
+		if err != nil {
+			return nil, err
+		}
+
+		off, err := strconv.ParseUint(ss[2], 16, 64)
+		if err != nil {
+			return nil, err
+		}
+
+		// Ignore errors and continue if not found.
+		note, _ := readSDTNote(path, provider, probe)
+		if note != nil {
+			note.path = path
+
+			if note.semaphore != 0 {
+				note.semaphore_offset_ptrace = note.semaphore + start - off
+			}
+
+			return note, nil
```

Comment:
Couldn't there be notes in multiple sections?

Reply:
there could be multiple notes in one section (`.note.stapsdt`) but every one of them has a different probe name or provider; since this is an example, we don't load them all but we search for the specific one we need and exit early if found